src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1194 @@
+                    }
+
+                    State newState = state;
+                    newState.SetIsLocked();
+
+                    State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
+                    if (stateBeforeUpdate == state)
+                    {
+                        return TryLockResult.Locked;
+                    }
+
+                    state = stateBeforeUpdate;
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool TryLockBeforeWait(Lock lockObj)
+            {
+                // This method is called before waiting. It must either acquire the lock or register a waiter. It also keeps
+                // track of the waiter starvation start time.
+
+                var state = new State(lockObj);
+                bool waiterStartTimeWasReset = false;
+                while (true)
+                {
+                    State newState = state;
+                    if (newState.ShouldNonWaiterAttemptToAcquireLock)
+                    {
+                        newState.SetIsLocked();
+                    }
+                    else
+                    {
+                        if (!newState.TryIncrementWaiterCount())
+                        {
+                            ThrowHelper.ThrowOutOfMemoryException_LockEnter_WaiterCountOverflow();
+                        }
+
+                        if (!state.HasAnyWaiters && !waiterStartTimeWasReset)
+                        {
+                            // This would be the first waiter. Once the waiter is registered, another thread may check the
+                            // waiter starvation start time and the previously recorded value may be stale, causing
+                            // ShouldNotPreemptWaitersMask to be set unnecessarily. Reset the start time before registering the
+                            // waiter.
+                            waiterStartTimeWasReset = true;
+                            lockObj.ResetWaiterStartTime();
+                        }
+                    }
+
+                    State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
+                    if (stateBeforeUpdate == state)
+                    {
+                        if (state.ShouldNonWaiterAttemptToAcquireLock)
+                        {
+                            return true;
+                        }
+
+                        Debug.Assert(state.HasAnyWaiters || waiterStartTimeWasReset);
+                        if (!state.HasAnyWaiters || waiterStartTimeWasReset)
+                        {
+                            // This was the first waiter or the waiter start time was reset, record the waiter start time
+                            lockObj.RecordWaiterStartTime();
+                        }
+                        return false;
+                    }
+
+                    state = stateBeforeUpdate;
```

Comment:
We have a lot of case of a pattern here like:
```cs
while (true)
{
     newState = DesiredUpdate(state);
     State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
     ...
     if (stateBeforeUpdate == state)
             return;
     ...
     state = stateBeforeUpdate;
}
```

Some of these could be less "hot" than others, hard to tell, but they all try to update the same shared location in a loop. If we have a lot of threads trying to change the state, being eager will just make everybody slower since the memory bus capacity is limited.
On the other hand every collision informs us that we have more threads trying to update the state than we thought, and it makes sense to introduce growing delays before retrying. 

```cs

int iteration = 0;
while (true)
{
     newState = DesiredUpdate(state);
     State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
     ...
     if (stateBeforeUpdate == state)
             return;
     ...
     ExponentialBackoff(iteration++);
     // read the new state. It will be cheap if did not change; if changed, we want the new state
     state = _state;
}
```

Reply:
Responded here: https://github.com/dotnet/runtime/pull/87672#discussion_r1237335803